### PR TITLE
http: changing mutate[request|response] to special case all upgrade headers

### PR DESCRIPTION
**Pull request by alyssawilk**
Extending the mutateRequestHeaders to special case all upgrades, not just WebSocket.

Adding special handling for mutateResponse headers to do similar special casing of upgrade responses.

Changing the utility we use to determine if there is an upgrade header.  The now removed caseInsensitiveContains had a bug where it would parse "upgrade: Websocket Keep-Alive" as a websocket upgrade.  The existing Envoy::StringUtil::caseFindToken will parse that as one token but still handles "upgrade: Webocket, keep-alive" correctly.

*Risk Level*: Medium (changing existing websocket handling)
*Testing*: new unit tests, regression test for fixed bug.
*Docs Changes*: n/a
*Release Notes*: not added.  We could note the bugfix if we think it noteworthy.
Part of #3301

### CODE REVIEW

**Comment by alyssawilk** [comment]
So I noticed that I was stripping the transferEncoding header on the response path and not the request path, which was clearly inconsistent and wrong. After some poking around and a question posited on #3301 I'm leaning towards just properly handling the transfer encoding header as I proposed there.

For now, I'm leaving both mutate functions as-is with respect to the T-E header but simply adding tests of the existing behavior and TODOs to fix it.

File: source/common/http/conn_manager_utility.cc [hunk 1]
```diff
@@ -303,7 +304,19 @@ void ConnectionManagerUtility::mutateXfccRequestHeader(Http::HeaderMap& request_
...
-  response_headers.removeConnection();
+  if (Utility::isUpgrade(request_headers) && Utility::isUpgrade(response_headers)) {
+    // As in mutateRequestHeaders, Upgrade responses have special handling.
+    //
+    // Unlike mutateRequestHeaders there is no explicit protocol check. If Envoy is proxying an
+    // upgrade response it has already passed the protocol checks.
+    const bool no_body =
+        (!response_headers.TransferEncoding() && !response_headers.ContentLength());
+    if (no_body) {
+      response_headers.insertContentLength().value(uint64_t(0));
```
Comment (mattklein123) [inline source/common/http/conn_manager_utility.cc]: Why is this needed in the upgrade path but not other cases? I think I'm just ignorant of the standard here...

Comment (alyssawilk) [reply source/common/http/conn_manager_utility.cc]: For response headers I believe you can frame by connection close.  That won't work for WebSocket upgrades because you can't close between HTTP and WebSocket payload.
